File: src/sorting/TopKFrequent.java
package sorting;

import java.util.Arrays;
import java.util.HashMap;
import java.util.PriorityQueue;

public class TopKFrequent {

    public static void main(String[] args) {

        int[] nums = new int[]{1,1,1,2,2,3};
        int k = 2;

        System.out.println("Answer"+Arrays.toString(topKFrequent(nums, k)));

    }

    public static int[] topKFrequent(int[] nums, int k) {

        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> Integer.compare(a[1], b[1]));

        // num , freq
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        for(int num: nums){
            map.put(num, map.getOrDefault(num, 0)+1);
        }



        for(int key : map.keySet()){
            int[] temp = new int[2];
            temp[0] = key;
            temp[1] = map.get(key);

            heap.add(temp);

            if (heap.size() > k) {
                int[] smthng = heap.poll();
                System.out.println("smthng="+Arrays.toString(smthng));
            }
        }

        int[] result = new int[heap.size()];

        int i = result.length-1;
        while(!heap.isEmpty()) {
            int[] tmp = heap.poll();
            result[i] = tmp[0];
            i--;
        }

        return result;
    }
}

File: src/sorting/TopKFrequent.kt
package sorting

import java.util.*

// link: https://leetcode.com/problems/top-k-frequent-elements/

fun main() {
    topKFrequent2(intArrayOf(1, 1, 1, 2, 2, 3), 2).forEach { print("$it ") }
    println("")
    topKFrequent2(intArrayOf(1, 2, 3, 4, 4), 2).forEach { print("$it ") }
}

fun topKFrequent(nums: IntArray, k: Int): IntArray {

    // key - num, value - frequency
    val countMap = mutableMapOf<Int, Int>()

    // max heap - max frequency at the top and sorted by frequency in descending order
    val heap = PriorityQueue<IntArray>(compareBy { it[1] })

    for (i in nums.indices) {
        countMap[nums[i]] = countMap.getOrDefault(nums[i], 0) + 1
    }

    println(countMap)

    for (key in countMap.keys) {
        val first = key
        val second = countMap[key]

        heap.add(intArrayOf(first, second!!))

        if (heap.size > k) {
            val polledValue = heap.poll()
            println("polledValue = ${polledValue!!.contentToString()}")
        }
    }

    //print("heap = $heap")
    val result = IntArray(k)
    var i = k - 1
    while (!heap.isEmpty()) {
        result[i] = heap.poll()[0]
        i--
    }

    return result
}


fun topKFrequent2(nums: IntArray, k: Int): IntArray {
    val numsMap = hashMapOf<Int, Int>()
    val priorityQueue = PriorityQueue<Pair<Int, Int>>(compareByDescending { it.second })

    nums.forEach { num ->
        numsMap[num] = numsMap.getOrDefault(num, 0) + 1
    }

    for ((key, value) in numsMap) {
        priorityQueue.add(Pair(key, value))
    }

    val res = IntArray(k)
    for (i in 0 until k) {
        res[i] = priorityQueue.poll().first
    }
    return res
}